pkg/signup/service/signup_service.go

```diff
@@ -161,6 +162,93 @@ func EncodeUserID(subject string) string {
+		return nil, err
+	}
+
+	var event *toolchainv1alpha1.ToolchainEvent
+	now := time.Now()
+
+	for _, e := range toolchainEvents.Items {
+		readyCondition, found := condition.FindConditionByType(e.Status.Conditions, toolchainv1alpha1.ToolchainEventReady)
+		// If the event is found and it is currently ready and active
+		if found && readyCondition.Status == apiv1.ConditionTrue && e.Spec.StartTime.Time.Before(now) && e.Spec.EndTime.Time.After(now) {
+			if event != nil {
+				// If an event has already been found, then it means there are multiple ToolchainEvent resources with the
+				// same activation code.  Return an error here
+				err := errors.NewInternalError(errors2.New("multiple matching ToolchainEvent resources found with duplicate activation code"))
+				log.Error(ctx, err, "multiple ToolchainEvent resources found")
+				return nil, err
+			}
+			evt := e
+			event = &evt
+		}
+	}
+
+	// If no event was found, return an error
+	if event == nil {
+		err := errors.NewBadRequest("invalid activation code")
+		log.Error(ctx, err, "invalid activation code")
+		return nil, err
+	}
+
+	// If the event has no more activations available, return an error
+	if event.Status.ActivationCount >= event.Spec.MaxAttendees {
+		err := errors.NewBadRequest("limit reached for activation code")
+		log.Error(ctx, err, "limit reached for activation code")
+		return nil, err
+	}
+
+	// First check if the user already exists, if they do and are currently active, return an error
+
+	encodedUserID := EncodeUserID(ctx.GetString(context.SubKey))
+
+	// Try to retrieve UserSignup resource from the host cluster
+	userSignup, err := s.CRTClient().V1Alpha1().UserSignups().Get(encodedUserID)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Increment the activation count for this event
+			err = s.incrementEventActivationCount(ctx, event)
+			if err != nil {
+				return nil, err
+			}
+			// New Signup
+			log.WithValues(map[string]interface{}{"encoded_user_id": encodedUserID}).Info(ctx, "user not found, creating a new one")
+			return s.createUserSignup(ctx, event)
+		}
+		return nil, err
+	}
+
+	// If the user does exist but they are deactivated, they may proceed with the activation
+	signupCondition, found := condition.FindConditionByType(userSignup.Status.Conditions, toolchainv1alpha1.UserSignupComplete)
+	if found && signupCondition.Status == apiv1.ConditionTrue && signupCondition.Reason == toolchainv1alpha1.UserSignupUserDeactivatedReason {
+		// Increment the activation count for this event
+		err = s.incrementEventActivationCount(ctx, event)
+		if err != nil {
+			return nil, err
+		}
+		// Signup is deactivated. We need to reactivate it
+		return s.reactivateUserSignup(ctx, userSignup, event)
+	}
+
+	return nil, nil
+}
+
+func (s *ServiceImpl) incrementEventActivationCount(ctx *gin.Context, event *toolchainv1alpha1.ToolchainEvent) error {
+	event.Status.ActivationCount++
+	_, err := s.CRTClient().V1Alpha1().ToolchainEvents().Update(event)
```

Comment:
This is a good point, especially if an activation code is simultaneously provided to a large number of attendees at the start of an event.  I will have to think a little about the best way to handle this, my first thought is to just implement it as a set number of retries, however I don't want to just jump in and code it this way without considering alternatives.

Reply:
whatever the solution that is implemented, it would be nice to have a few stress tests in our e2e repo to verify that the implementation handles concurrent signups